addons/point_of_sale/tests/test_pos_stock_account.py

```diff
@@ -235,3 +235,312 @@ def test_03_order_product_w_owner(self):
+            'line_ids': [
+                {'account_id': self.sales_account.id, 'debit': 0, 'credit': 30, 'reconciled': False},
+                {'account_id': self.expense_account.id, 'debit': 15, 'credit': 0, 'reconciled': False},
+                {'account_id': self.cash_pm1.receivable_account_id.id, 'debit': 30, 'credit': 0, 'reconciled': True},
+                {'account_id': self.output_account.id, 'debit': 0, 'credit': 15, 'reconciled': True},
+            ],
+        })
+
+        # Verify the expected inventory valuation entry was created
+        stock_moves = self.pos_session.picking_ids[0].move_lines.account_move_ids
+        self.assertEqual(len(stock_moves), 1)
+        self._assert_account_move(stock_moves[0], {
+            'line_ids': [
+                {'account_id': self.valuation_account.id, 'debit': 0, 'credit': 15, 'reconciled': False},
+                {'account_id': self.output_account.id, 'debit': 15, 'credit': 0, 'reconciled': True},
+            ],
+        })
+
+    def test_selling_a_product_with_an_invalid_picking(self):
+        self.open_new_session()
+
+        # create orders
+        orders = [self.create_ui_order_data([(self.product1, 1)])]
+
+        # sync orders
+        self.env['pos.order'].create_from_ui(orders)
+
+        # close the session
+        self.pos_session.action_pos_session_validate()
+
+        # picking should be in confirmed state
+        self.assertEqual(self.pos_session.picking_ids[0].state, 'confirmed', 'Picking should be in confirmed state.')
+
+        # The expense and stock output lines will be 0 because the stock picking failed. The stock output line won't
+        # be reconciled.
+        self._assert_account_move(self.pos_session.move_id, {
+            'line_ids': [
+                {'account_id': self.sales_account.id, 'debit': 0, 'credit': 20, 'reconciled': False},
+                {'account_id': self.expense_account.id, 'debit': 0, 'credit': 0, 'reconciled': False},
+                {'account_id': self.cash_pm1.receivable_account_id.id, 'debit': 20, 'credit': 0, 'reconciled': True},
+                {'account_id': self.output_account.id, 'debit': 0, 'credit': 0, 'reconciled': False},
+            ],
+        })
+
+        # Set a serial number for the product
+        lot = self.env['stock.production.lot'].create({
+            'name': 'lot1',
+            'product_id': self.product1.id,
+            'company_id': self.env.company.id,
+        })
+
+        self.pos_session.picking_ids[0].move_line_ids[0].write({
+            'lot_id': lot.id
+        })
+
+        # Validate the stock picking
+        self.pos_session.picking_ids[0].button_validate()
+
+        # Verify the expected stock moves were created
+        stock_moves = self.pos_session.picking_ids[0].move_lines.account_move_ids
+        self.assertEqual(len(stock_moves), 2)
+        # Sort the moves in the order they were created in
+        stock_moves = sorted(stock_moves, key=lambda x: x.name)
```

Comment:
I don't have a better name in mind but `stock_moves` is too generic.
It could either be an `account.move` or a `stock.move`

Reply:
Also, the sorting for `account.move` is `_order = 'date desc, name desc, id desc'`.
Thus, you just need to reverse them with `stock_moves = stock_moves[::-1]`.

The performance probably doesn't really matter but I expected that to be much faster than using `sorted`

Reply:
Agreed on the name, I'll try to be more specific. Regarding the default order of the `account_move_ids`: it's different between Odoo 14.0 and Odoo 15.0. That's why I'm sorting on the name, so the test doesn't depend on that implementation detail. Otherwise I would be checking `stock_moves[0]` in one version and `stock_moves[1]` in the other one, which seems a bit too fragile to me.

Reply:
I'm going for `stock_picking_account_moves` instead of `stock_moves`. A bit uglier, but definitely more specific :-)